sklearn/neighbors/lmnn.py

```diff
@@ -0,0 +1,1122 @@
+        self.n_iter_ = 0
+        opt_result = minimize(**optimizer_params)
+
+        # Reshape the solution found by the optimizer
+        self.transformation_ = opt_result.x.reshape(-1, X_valid.shape[1])
+
+        # Stop timer
+        t_train = time.time() - t_train
+        if self.verbose:
+            print('[{}] Training took {:8.2f}s.'.format(
+                self.__class__.__name__, t_train))
+
+        # Optionally store information returned by the optimizer
+        if self.store_opt_result:
+            self.opt_result_ = opt_result
+
+        return self
+
+    def transform(self, X):
+        """Applies the learned transformation to the given data.
+
+        Parameters
+        ----------
+        X : array-like, shape (n_samples, n_features)
+            Data samples.
+
+        Returns
+        -------
+        X_embedded: array, shape (n_samples, n_features_out)
+            The data samples transformed.
+
+        Raises
+        ------
+        NotFittedError
+            If :meth:`fit` has not been called before.
+        """
+
+        check_is_fitted(self, ['transformation_'])
+        X = check_array(X)
+
+        return np.dot(X, self.transformation_.T)
+
+    def _transform_without_checks(self, X):
+        """Same as transform but without validating the inputs.
+
+        Parameters
+        ----------
+        X : array, shape (n_samples, n_features)
+            Data samples.
+
+        Returns
+        -------
+        X_embedded: array, shape (n_samples, n_features_out)
+            The data samples transformed.
+
+        """
+        return np.dot(X, self.transformation_.T)
+
+    def _validate_params(self, X, y):
+        """Validate parameters as soon as :meth:`fit` is called.
+
+        Parameters
+        ----------
+        X : array-like, shape (n_samples, n_features)
+            The training samples.
+
+        y : array-like, shape (n_samples,)
+            The corresponding training labels.
+
+        Returns
+        -------
+        X : array, shape (n_samples, n_features)
+            The validated training samples.
+
+        y_inverse : array, shape (n_samples,)
+            The validated training labels, encoded to be integers in
+            the range(0, n_classes).
+
+        init : string or numpy array
+            The validated initialization of the linear transformation.
+
+        Raises
+        -------
+        TypeError
+            If a parameter's type does not match the desired type.
+
+        ValueError
+            If a parameter's value violates its legal value range or if the
+            combination of two or more given parameters is incompatible.
+        """
+
+        # Find the appearing classes and the class index for each sample
+        classes, y_inverse = np.unique(y, return_inverse=True)
+        classes_inverse = np.arange(len(classes))
+
+        # Ignore classes that have less than 2 samples (singleton classes)
+        class_sizes = np.bincount(y_inverse)
```

Comment:
maybe it's time we backported `unique(..., return_counts)` into `sklearn.utils.fixes` to take advantage of their implementation of unique + bincount.

Reply:
Sounds good, except if dependency to numpy is going to be bumped to 1.9.

Reply:
I don't think that'll happen so soon.